### DISCUSSION TITLE
Composite foreign key

### DISCUSSION THREAD

**Discussion by piotr-pawlowski** [mikro-orm/mikro-orm#1952] [Q&A]
Hi! Could you show an example how to define a relation between two entities (eg. one-to-many, many-to-one) based on two columns (composite fk)? I can't find that in doc.

**Comment by B4nan** [comment]
What have you tried?

**Reply by piotr-pawlowski** [reply]
I have tree entities: `EntityA`, `EntityB`, `EntityC` with given relations many-to-many: A-B, B-C. And I need a relation A-B-C (many-to-many), but this relation might exist only if exists relation between A-B and B-C.

```
@Entity()
export class ABC {

  @ManyToOne(() => EntityA, { primary: true, name: 'aId' })
  entityA: EntityA

  @ManyToOne(() => EntityB, { primary: true, name: 'bId' })
  entityB: EntityB

  @ManyToOne(() => EntityC, { primary: true, name: 'cId' })
  entityC: EntityC

  @ManyToOne(() => AtoB)
  aToB: AtoB

  @ManyToOne(() => BtoC)
  bToC: BtoC

  [PrimaryKeyType]: [number, number, number]
}
```
`AtoB` is many-to-many table with `aId` and `bId`. `BtoC` is many-to-many table with `bId` and `cId`.

Unfortunately the code above generates table `ABC` with 7 columns (3 columns for primary key) and 4 columns with two composite fks. I would like to achive solution that uses primary key to be foreign key as well (`aId` and `bId` should reference to the `AtoB` and `bId` with `cId` should reference to thie `BtoC`).

**Reply by B4nan** [reply]
You already have 3 FKs as PKs, but they apparently point to different entity than `AtoB/BtoC`, so I don't really understand what you are up to. Maybe you want to use `AtoB/BtoC` as primary keys instead? You are talking about m:n but define m:1 - m:1 relation is always owner, and it will always produce the FK column(s).

**Reply by piotr-pawlowski** [reply]
Ok, you're right. I have made some changes:
```
@Entity()
export class ABC {

  @ManyToOne(() => AtoB, { primary: true })
  aToB: AtoB

  @ManyToOne(() => BtoC, { primary: true })
  bToC: BtoC

  [PrimaryKeyType]: [number, number, number]
}
```
The code above generates composite primary key that consists of 4 columns...

Case 2:
 ```
@Entity()
export class ABC {

  @ManyToOne(() => AtoB, { primary: true, joinColumns: ['aId', 'bId'] })
  aToB: AtoB

  @ManyToOne(() => BtoC, { primary: true, joinColumns: ['bId', 'cId'] })
  bToC: BtoC

  [PrimaryKeyType]: [number, number, number]
}
```
In case above column `bId` is duplicated in create statement so I can't create schema.

**Reply by B4nan** [reply]
Well, you are explicitly saying how the columns should be named (via `joinColumns`), so just change it.

**Reply by B4nan** [reply]
For the 2 relations you will always have to have 4 columns. You can't just reuse the `bId` column in both, that does not make any sense - it would mean there is just one relation, not two.

**Reply by piotr-pawlowski** [reply]
> You can't just reuse the bId column in both, that does not make any sense

I can't do that in MikroORM, but in SQL is best practise. Why do you think it doesn't make sense? What is wrong with this schema?

```
CREATE TABLE `ABC` (
  `aId` int(11) NOT NULL,
  `bId` int(11) NOT NULL,
  `cId` int(11) NOT NULL,
 
  PRIMARY KEY (`aId`, `bId`, `cId`),
 
  CONSTRAINT `FK_AB` FOREIGN KEY (`aId`, `bId`) REFERENCES `AtoB` (`aId`, `bId`) ON DELETE CASCADE ON UPDATE CASCADE,
  CONSTRAINT `FK_BC` FOREIGN KEY (`bId`, `cId`) REFERENCES `BtoC` (`bId`, `cId`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB
```

**Reply by B4nan** [reply]
> I can't do that in MikroORM, but in SQL is best practise. 

You can do it, but not this way. This way it does not make any sense, as you would introduce inconsistency to the PK (one of the PKs could have different `bId` than the other).

**Reply by B4nan** [reply]
Maybe give me the schema you want as a result, so I can see what you are up to.

**Reply by piotr-pawlowski** [reply]
> This way it does not make any sense, as you would introduce inconsistency to the PK (one of the PKs could have different bId than the other).

What do you mean? There are constraints that do not allowed to introduce inconsistency. There is only one column `bId` in that table. 

```
CREATE TABLE `A` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB;

CREATE TABLE `B` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB;

CREATE TABLE `C` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB;

CREATE TABLE `AtoB` (
  `aId` int(11) NOT NULL,
  `bId` int(11) NOT NULL,
  PRIMARY KEY (`aId`,`bId`),
  CONSTRAINT `FK_AtoB_A` FOREIGN KEY (`aId`) REFERENCES `A` (`id`) ON UPDATE CASCADE,
  CONSTRAINT `FK_Ato_B_B` FOREIGN KEY (`bId`) REFERENCES `B` (`id`) ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;

CREATE TABLE `BtoC` (
  `bId` int(11) NOT NULL,
  `cId` int(11) NOT NULL,
  PRIMARY KEY (`bId`,`cId`),
  CONSTRAINT `FK_BtoC_B` FOREIGN KEY (`bId`) REFERENCES `B` (`id`) ON UPDATE CASCADE,
  CONSTRAINT `FK_BtoC_C` FOREIGN KEY (`cId`) REFERENCES `C` (`id`) ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;

CREATE TABLE `ABC` (
  `aId` int(11) NOT NULL,
  `bId` int(11) NOT NULL,
  `cId` int(11) NOT NULL,
 
  PRIMARY KEY (`aId`, `bId`, `cId`),
 
  CONSTRAINT `FK_ABC_AB` FOREIGN KEY (`aId`, `bId`) REFERENCES `AtoB` (`aId`, `bId`) ON DELETE CASCADE ON UPDATE CASCADE,
  CONSTRAINT `FK_ABC_BC` FOREIGN KEY (`bId`, `cId`) REFERENCES `BtoC` (`bId`, `cId`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB
```
![image](https://user-images.githubusercontent.com/33734186/122767928-6ae7a400-d2a3-11eb-8ef0-9b040a963730.png)

I can't share full schema with you because it is commercial project.

**Reply by B4nan** [reply]
> What do you mean? There are constraints that do not allowed to introduce inconsistency.

In your table schema, yes, but not in your entity definition. You define two relations, both have the `b` column, but they are not bound to each other anyhow - you can set two different `b` columns via your FKs.

You are probably right, I really don't think you can achieve this schema with MikroORM, and I really don't think this is the right way to model it. Sharing one column in two FKs is not supported and seems like a bad design to me (on the other hand this is valid schema, so it should be possible to map it somehow). 

Do you know how other similar ORMs handle this case? I guess what you have in `case 2` is how it should look like, but we need to say which one should be ignored in the schema definition. Maybe it could work automatically, if we detect two exact same column definitions. I'd rather keep such change for v5, but after giving it some thoughts, it feels ok/safe to me.

**Reply by piotr-pawlowski** [reply]
It is typical diamond shape design. I know it cannot be achived with Sequelize because it doesn't support composite foreign key (known as 311 issue). That's why I am looking for new ORM that I can migrate from Sequelize. I am going to check TypeORM but it has other known bugs that I am afraid of.

**Reply by piotr-pawlowski** [reply]
Any updates on this issue?
